cmd/gen-go-sample/main.go

```diff
@@ -79,32 +93,97 @@ func main() {
 		}
 
 		gen.descInfo = pbinfo.Of(gen.desc.GetFile())
-		donec <- struct{}{}
 	}()
 
-	// TODO(hzyi): Read sample config
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := readSampleConfigs(&gen, *samplePath); err != nil {
+			log.Fatal(err)
+		}
+	}()
 
 	if err := os.MkdirAll(*outDir, 0755); err != nil {
 		log.Fatal(err)
 	}
 
-	<-donec
-	<-donec
+	wg.Wait()
 
-	if err := genMethodSamples(&gen, gen.sampleConfig, *nofmt, *outDir); err != nil {
+	if err := genMethodSamples(&gen, *nofmt, *outDir); err != nil {
 		log.Fatal(err)
 	}
 
 }
 
-func genMethodSamples(gen *generator, sampConf schema_v1p2.SampleConfig, nofmt bool, outDir string) error {
-	for _, samp := range sampConf.Samples {
+func readSampleConfigs(gen *generator, path string) error {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return errors.E(err, "cannot read sample config files: %s", path)
+	}
+	switch mode := fi.Mode(); {
+	case mode.IsDir():
+		var sc schema_v1p2.SampleConfig
+		err := filepath.Walk(path,
+			func(p string, info os.FileInfo, err error) error {
+				if err != nil {
+					return errors.E(err, "cannot read sample config file: %s", p)
+				}
+				if info.IsDir() || !strings.HasSuffix(p, ".yaml") {
+					// ignore directories and non-YAML files
+					return nil
+				}
+
+				f, err := os.Open(p)
+				if err != nil {
+					return errors.E(err, "cannot read sample config file: %s", p)
+				}
+
+				if err := yaml.NewDecoder(f).Decode(&sc); err != nil {
+					// ignore unrecognized YAML files
+					return nil
+				}
+				if sc.Type != expectedSampleConfigType {
+					// ignore non sample config files
+					return nil
+				}
+				if sc.Version != expectedSampleConfigVersion {
+					// ignore unsupported versions
+					return nil
+				}
+				gen.sampleConfig.Samples = append(gen.sampleConfig.Samples, sc.Samples...)
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+
+	case mode.IsRegular():
+		f, err := os.Open(path)
+		if err != nil {
+			return errors.E(err, "cannot read sample config file: %s", path)
+		}
+		if err := yaml.NewDecoder(f).Decode(&gen.sampleConfig); err != nil {
```

Comment:
Support multiple YAML documents in the file and error if none of them are a sample config.

Reply:
Ack

Reply:
Done.